### ISSUE TITLE
BUG: np.array(x) converts np.nan to string, if an element in x is a string

### ISSUE THREAD

**Issue by Jakobhenningjensen** [numpy/numpy#22042]
### Describe the issue:

The array `X = np.array([np.nan,np.nan,"foo",np.nan,np.nan])` converts the `np.nan` to `'nan'` i.e a string instead of keeping it as `np.nan`. This can really easy introduce bugs, since stuff like `pd.issnull()` now evaluates the `'nan'` to `False` instead of `True`.

This behaviour/conversion might be intentional (since `type(np.nan)==float`) but if it is I think it should either raise a warning since the conversion happens behind the scenes, or should be controlled by e.g a `conversion` argument which does not allow to convert the elements to other types, or add it more clearly in the documentation.

Eitherway I think it's a very dangerous behaviour as it is right now.

### Reproduce the code example:

```python
import numpy as np
import pan
X = [np.nan,np.nan,"foo",np.nan,np.nan]
idx_to_keep =pd.notnull(X) # [False,False,True,False,False]
.
.
X = np.array(X)
idx_to_keep =pd.notnull(X) # [False,False,False,False,False]
```

### NumPy/Python version information:

1.22.4 3.10.2 (tags/v3.10.2:a58ebcc, Jan 17 2022, 14:12:15) [MSC v.1929 64 bit (AMD64)]

**Comment by BrentTanKian** [comment]
Hey I tested this out and do agree that this is misleading behaviour. I would like to work on this, and was wondering if converting to empty string instead of 'nan' is the best way forward? 

Was also wondering if you might have had a typo for the result of doing pd.notnull() on X after you converted X to a numpy array because I got `[ True,  True,  True,  True,  True]`, instead of `[False,False,False,False,False]` like what you mentioned

**Comment by Jakobhenningjensen** [comment]
It's a good question - I think that some sort of "null-object" like empty string might be a good idea, since it sort of corresponds to np.nan.

Yes, it was indeed a typo; I have editted that

**Comment by BrentTanKian** [comment]
Yup that sounds good to me too. Would someone need to confirm on this, or could I proceed to push up some code that contains this new change?

**Comment by Jakobhenningjensen** [comment]
There might be an issue with  `np.isnan` (since it cannot evaluate `""`) and `pd.isnull("")` evaluates to `True`.

I don't know if that is something that should be thought of, or if it's up to the user to handle that.

The issue is that the `np.nan` suddenly changes it "characteristics" when converted to `""` instead (some sort) of `np.nan`.
I do not have a better suggestion than `""` though unless you wan't to introduce the string version of `np.nan` ie. `np.nas`?  But I do see some issues with empty string (if does make it more easy to understand the conversion though instead of just the `"nan"`)

**Comment by BrentTanKian** [comment]
I see, you're right about the empty string leading to possible future complications with other functions. Would you think that there would be some way to retain the original characteristics of np.nan (as a null value) instead of converting it to a string, after conversion to a numpy array? And also np.nas doesn't seem to be valid to me after experimenting with it for a bit on my end

**Comment by Jakobhenningjensen** [comment]
I must admit, that I'm not that deep into the "internal" of numpy arrays i.e what is doable and what is not.

The best thing would to keep it as a `np.nan`, but I assume that isn't doable since all objects must have same type?

**Comment by BrentTanKian** [comment]
Yeah I believe that is correct. Numpy arrays can only contain objects of a single type, which means a conversion will have to take place. Seems like the conversion is also dynamic according to the elements present in the original list. If "foo" was replaced with an integer, then the np.nan values get converted to float nan values, which still correctly show up as null values when pd.notnull() is used. So I guess really the only problem is with string conversion, and for this case, perhaps 'nas' might be better than an empty string?

**Comment by rkern** [comment]
`np.array()` has a collection of heuristics for determining what dtype to use when the objects are mixed. Typically, this just dealing with just integers mixed in with integers, upcasting integers to floats. But `np.array()` has to deal with all kinds of mixtures. In particular, when `np.array()` finds a string mixed in with numbers, it will "upcast" the numbers to strings by using `str()`, hence `np.nan` gets turned into `'nan'`. `pandas.Series()` uses a different set of rules since it doesn't have a string dtype. When it encounters a string, it will upcast everything to `dtype=object`.

If that's the behavior you want, be explicit by using `np.array(X, dtype=object)`. I don't think we will change how we convert `np.nan` to strings as a special case.

**Comment by seberg** [comment]
The issue is that we support promotion of float to string.  I tried to deprecate that once, but there was some fallout (not sure anymore what exactly, but for once, it was when we added `np.concatenate(..., dtype=)`).

It may be time to revisit such a deprecation, independend of this issue allowing promotion (rather than `dtype="U"`) is indeed strange.  OTOH, I am not sure we would allow it to go to `object` magically, rather you may have to pass `dtype=object` manually then.  Which is already the clean thing to do.

**Comment by seberg** [comment]
Ah, there are a couple of issues still open, that happened when I tried it: gh-18425, gh-19078, and maybe more.  The PR was gh-18116.

**Comment by Jakobhenningjensen** [comment]
@rkern that makes sense. But I would strongly advice having such automatical value-editting listed some where clearly. I don't think many would expect that a `np.nan` value suddenly turns into the string with the value `'nan'`. I would expect one of 2 things to happen:

1)  "TypeError: the value `np.nan` cannot be converted to a string"
2)  "Warning: the value `np.nan` has been converted to the value `'nan'`.

and maybe be able to set one of them as a parameter `convert_silently='error'`, `convert_silently='warning'` (default) or `convert_silently=True`.

**Comment by rkern** [comment]
`str(np.nan)` returns `'nan'` without error or warning, just like it does for any other `float` value. Why would we raise or warn in this context where `str()` is used to convert every value?

**Comment by seberg** [comment]
:+1: There is nothing special about `NaN`.  Special casing `NaN` only makes sense in the `pandas` world where `NaN` is used as a generic replacement for `NA`.

The argument that any numeric value should not suddenly be converted to a string without the user requesting it *can* be made, but this is not specific to `NaN` at all.

The desired behaviour would be that:
1. `np.array(["1", 1])` always fails, because there is no obvious *promotion* rule that the combination of strings and numbers should return strings.  Compare with float + int where we have fairly clear promotion rules.
2. `np.array(["1", 1, np.nan], dtype="U")` succeeds as it does now.  There is nothing special about converting to string (including the `NaN`).

**Comment by kdeshpande3** [comment]
Hi everyone, I'm interested in contributing to this issue as well. I understand the problem is with the conversion of np.nan to a string 'nan' when mixed data types are present in a NumPy array. I'd like to learn more about how to contribute and how I should start. Since it looks like this issue has been open for a while, are there any recommendations on tackling this issue? Thank you!

**Comment by rkern** [comment]
This is probably not a good starter issue. Changing the type promotion behavior is a policy decision rather than a technical code fix.

**Comment by kdeshpande3** [comment]
Thanks for letting me know. How can I determine if an issue is a good starter issue?
